### PR TITLE
xds/resolver: Optimize Interceptor Chain Construction

### PR DESCRIPTION
**Pull request by easwars**
#### Existing behavior:
- At routing time, when an RPC matches a route and a cluster is selected, the interceptor chain for that specific RPC is built.
- This chain is built on a per-RPC basis.
- A subsequent RPC that matches the exact same route and cluster will trigger the entire chain reconstruction again, even if no configuration has changed.

#### New  behavior:
- The interceptor chain is now pre-built for every route and every pickable cluster associated with that route.
- The chains are constructed once when the config selector is built.

#### Other changes:
- Existing unit tests have been converted to be more e2e style tests.
- This lays the necessary groundwork for upcoming changes to the filter API, specifically to support filter state retention

RELEASE NOTES: NONE

### CODE REVIEW

**Comment by easwars** [comment]
@eshitachandwani : FYI this might conflict with your resolver changes for A74. And since you have been looking at the resolver code for sometime now, it would be a good PR to review.

**Comment by dfawley** [comment]
@arjan-bal could you review this change, please?

File: internal/xds/resolver/serviceconfig.go [hunk 1]
```diff
@@ -363,6 +324,39 @@ func (cs *configSelector) stop() {
 	}
 }
 
+// newInterceptor builds a chain of client interceptors for the given filters.
+// The filter config override maps contain overrides from the route, cluster,
+// and virtual host respectively. The cluster override has the highest priority,
+// followed by the route override, and finally the virtual host override.
+func newInterceptor(filters []xdsresource.HTTPFilter, cluster, route, virtualHost map[string]httpfilter.FilterConfig) (iresolver.ClientInterceptor, error) {
```
Comment (arjan-bal) [inline internal/xds/resolver/serviceconfig.go]: nit: What do you think about adding an `override` suffix to `cluster`, `route`, and `virtualHost`? I find it can help with readability, but no worries if you prefer the current naming.

File: internal/xds/resolver/xds_resolver.go [hunk 2]
```diff
@@ -446,7 +446,10 @@ func (r *xdsResolver) onResolutionComplete() {
 		return
 	}
 
-	cs := r.newConfigSelector()
+	cs, err := r.newConfigSelector()
+	if err != nil {
+		r.onResourceError(fmt.Errorf("xds: failed to create config selector: %v", err))
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_resolver.go:452]: Should the function return early after calling `onResourceError`? I'm not sure what happens when a `nil` config selector is passed to `sendNewServiceConfig`.

**Review by arjan-bal — COMMENTED** [review]
Leaving my comments on the non-test code, still reviewing the tests.

File: internal/xds/resolver/serviceconfig.go [hunk 3]
```diff
@@ -363,6 +324,39 @@ func (cs *configSelector) stop() {
...
+		if override == nil {
+			override = route[filter.Name]
+		}
+		if override == nil {
+			override = virtualHost[filter.Name]
+		}
+		ib, ok := filter.Filter.(httpfilter.ClientInterceptorBuilder)
+		if !ok {
+			// Should not happen if it passed xdsClient validation.
+			panic(fmt.Sprintf("filter %q does not support use in client", filter.Name))
```
Comment (arjan-bal) [inline internal/xds/resolver/serviceconfig.go]: What do you think about continuing to return an error here? The method signature already allows for it, so callers presumably have error handling in place. Panicking to crash the process feels like a worse alternative, especially if this is a recoverable error.

File: internal/xds/resolver/xds_http_filters_test.go [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+
+func (*testHTTPFilterWithRPCMetadata) IsTerminal() bool { return false }
+
+var _ httpfilter.ClientInterceptorBuilder = &testHTTPFilterWithRPCMetadata{}
+
+func (fb *testHTTPFilterWithRPCMetadata) BuildClientInterceptor(config, override httpfilter.FilterConfig) (iresolver.ClientInterceptor, error) {
+	fb.logger.Logf("BuildClientInterceptor called with config: %+v, override: %+v", config, override)
+
+	if config == nil {
+		panic("unexpected missing config")
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go]: In my opinion printing an error to fail the test may be preferred to panicking from a spawned goroutine. The situation seems similar to the style guide recommendation about only calling `t.Fatalf` from the main test function.

File: internal/xds/resolver/xds_http_filters_test.go [hunk 5]
```diff
@@ -0,0 +1,678 @@
...
+// testHTTPFilterWithRPCMetadata is a HTTP filter used for testing purposes.
+//
+// This filter is used to verify that the xDS resolver and filter stack
+// correctly propagate filter configuration (both base and override) to RPCs. It
+// does this by injecting the config paths from its base and override configs as
+// JSON-encoded metadata into outgoing RPCs.  The metadata can then be observed
+// by the backend, allowing tests to assert that the correct filter
+// configuration was applied for each RPC.
+type testHTTPFilterWithRPCMetadata struct {
+	logger        logger
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go:108]: nit: Is there a benefit of using a `logger` interface here? If not, I think it's simpler to directory store a `testing.T` here .

File: internal/xds/resolver/xds_http_filters_test.go [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+	return filterConfigFromProto(cfg)
+}
+
+func (*testHTTPFilterWithRPCMetadata) ParseFilterConfigOverride(override proto.Message) (httpfilter.FilterConfig, error) {
+	return filterConfigFromProto(override)
+}
+
+func (*testHTTPFilterWithRPCMetadata) IsTerminal() bool { return false }
+
+var _ httpfilter.ClientInterceptorBuilder = &testHTTPFilterWithRPCMetadata{}
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go:127]: nit: I think we should add a comment here stating that `ClientInterceptorBuilder` is an optional interface for `Filter`s to implement so this compile time check ensures the test filter implements it. In my opinion this would help readers.

File: internal/xds/resolver/xds_http_filters_test.go [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+	if override != nil {
+		overrideCfg := override.(testFilterCfg)
+		overridePath = overrideCfg.path
+		if overrideCfg.newStreamErr != nil {
+			newStreamErr = overrideCfg.newStreamErr
+		}
+	}
+	var errStr string
+	if newStreamErr != nil {
+		errStr = newStreamErr.Error()
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go]: I see that we're converting a string to an error, assigning it to `testFilterCfg.newStreamErr` and converting it back to a string here. Can the error string be stored in `testFilterCfg.newStreamErr` to avoid these conversions?

File: internal/xds/resolver/xds_http_filters_test.go [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+	return &clientStream{ClientStream: cs, ctx: ctx}, nil
+}
+
+type clientStream struct {
+	iresolver.ClientStream
+	ctx context.Context
+}
+
+func (cs *clientStream) Context() context.Context {
+	return cs.ctx
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go]: Do we need to store a context here? Can we return `cs.ClientStream.Context()` instead?

File: internal/xds/resolver/xds_http_filters_test.go [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{AllowResourceSubset: true})
+	defer mgmtServer.Stop()
+
+	// Create an xDS resolver with bootstrap configuration pointing to the above
+	// management server.
+	nodeID := uuid.New().String()
+	bootstrapContents := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	if internal.NewXDSResolverWithConfigForTesting == nil {
+		t.Fatalf("internal.NewXDSResolverWithConfigForTesting is nil")
+	}
```
Comment (arjan-bal) [inline internal/xds/resolver/xds_http_filters_test.go:273]: nit: We can avoid this check and let the code panic if we don't expect to fail in most cases.

File: internal/xds/resolver/xds_resolver_test.go
Comment (arjan-bal) [inline internal/xds/resolver/xds_resolver_test.go:1]: There is a bunch of commented code here that probably needs to be removed. Can you please take a look?

File: internal/xds/resolver/xds_resolver.go (hunk 2 above)
Comment (eshitachandwani) [reply internal/xds/resolver/xds_resolver.go:452]: I think we should return early since `onResourceError` will call `sendNewServiceConfig` with an erroring config selector, so we should not call `sendNewServiceConfig` again with a nil config selector (which will set a default config selector in client conn)

File: internal/xds/resolver/serviceconfig.go (hunk 1 above)
Comment (easwars) [reply internal/xds/resolver/serviceconfig.go]: Done.

File: internal/xds/resolver/serviceconfig.go (hunk 3 above)
Comment (easwars) [reply internal/xds/resolver/serviceconfig.go]: At some point in the recent past, I got comfortable using panics for cases where we definitely don't expect something to happen, without some really bad programming error. This I believe is one such. But since an error is already being handled by the caller of this function, I'm ok to return an error here instead of panicking. Thanks.

File: internal/xds/resolver/xds_http_filters_test.go (hunk 5 above)
Comment (easwars) [reply internal/xds/resolver/xds_http_filters_test.go:108]: It's probably an overkill, but I feel that it's better to do this instead of accepting a `testing.T`, since we can very easily start calling other methods on it like `Error` or `Fatal` if we have access to it.

File: internal/xds/resolver/xds_http_filters_test.go (hunk 6 above)
Comment (easwars) [reply internal/xds/resolver/xds_http_filters_test.go:127]: Done.

File: internal/xds/resolver/xds_http_filters_test.go (hunk 4 above)
Comment (easwars) [reply internal/xds/resolver/xds_http_filters_test.go]: Done.

File: internal/xds/resolver/xds_http_filters_test.go (hunk 7 above)
Comment (easwars) [reply internal/xds/resolver/xds_http_filters_test.go]: Done.

File: internal/xds/resolver/xds_http_filters_test.go (hunk 8 above)
Comment (easwars) [reply internal/xds/resolver/xds_http_filters_test.go]: Looks like I don't need this type at all. I was probably doing something else at some point in time that required this type. But with the current state of the test and the current implementation of `testFilterInterceptor`, we don't need this type at all. Thanks.

File: internal/xds/resolver/xds_resolver_test.go
Comment (easwars) [reply internal/xds/resolver/xds_resolver_test.go:1]: Done. Thanks.

File: internal/xds/resolver/xds_resolver.go (hunk 2 above)
Comment (easwars) [reply internal/xds/resolver/xds_resolver.go:452]: Thanks for spotting this. Fixed it. 

Looks like we don't have any tests for this case, but this is also a very very unlikey case. There are two cases when building a config selector can fail:
- A filter is part of the configuration, but is not supported on the client side. This case would not pass xDS client validation. (I had a panic for this case initially, but changed it to return an error as part of this reivew).
- The call to build the interceptor for the filter fails. This is also very unlikey because if the filter successfully parsed the configuration (during resource parsing in the xDS client), there is very little reason for it to fail when trying to build the interceptor.

I'll anyway file an issue to add a test for this edge case.

**Comment by easwars** [comment]
@arjan-bal : Gentle ping on this one. This is blocking a design change that I need to make to support A83.

**Review by arjan-bal — APPROVED** [review]
LGTM!
